### PR TITLE
IRMutator->IRMutator2, Part 7

### PR DESCRIPTION
**Pull request by steven-johnson**

### CODE REVIEW

File: src/RemoveUndef.cpp [hunk 1]
```diff
@@ -106,52 +104,54 @@ class RemoveUndef : public IRMutator {
...
-            expr = Select::make(cond, t, f);
+            return Select::make(cond, t, f);
         }
     }
 
-    void visit(const Load *op) {
+    Expr visit(const Load *op) override {
         Expr pred = mutate(op->predicate);
-        if (!pred.defined()) return;
+        if (!pred.defined()) return op;
```
Comment (abadams) [inline src/RemoveUndef.cpp]: I think these should all be return Expr(), not return op. In this mutator, if you get an undefined Expr() back, you're supposed to pass it upwards as undefined.

File: src/RemoveUndef.cpp [hunk 2]
```diff
@@ -106,52 +104,54 @@ class RemoveUndef : public IRMutator {
...
         Expr index = mutate(op->index);
-        if (!index.defined()) return;
+        if (!index.defined()) return op;
```
Comment (abadams) [inline src/RemoveUndef.cpp]: return Expr()

File: src/RemoveUndef.cpp [hunk 3]
```diff
@@ -106,52 +104,54 @@ class RemoveUndef : public IRMutator {
...
-            expr = Load::make(op->type, op->name, index, op->image, op->param, pred);
+            return Load::make(op->type, op->name, index, op->image, op->param, pred);
         }
     }
 
-    void visit(const Ramp *op) {
+    Expr visit(const Ramp *op) override {
         Expr base = mutate(op->base);
-        if (!base.defined()) return;
+        if (!base.defined()) return op;
```
Comment (abadams) [inline src/RemoveUndef.cpp]: return Expr()

File: src/RemoveUndef.cpp [hunk 4]
```diff
@@ -106,52 +104,54 @@ class RemoveUndef : public IRMutator {
...
         Expr stride = mutate(op->stride);
-        if (!stride.defined()) return;
+        if (!stride.defined()) return op;
```
Comment (abadams) [inline src/RemoveUndef.cpp]: return Expr()

File: src/RemoveUndef.cpp [hunk 5]
```diff
@@ -161,20 +161,20 @@ class RemoveUndef : public IRMutator {
         for (size_t i = 0; i < op->args.size(); i++) {
             Expr old_arg = op->args[i];
             Expr new_arg = mutate(old_arg);
-            if (!new_arg.defined()) return;
+            if (!new_arg.defined()) return op;
```
Comment (abadams) [inline src/RemoveUndef.cpp]: return Expr()

File: src/RemoveUndef.cpp [hunk 6]
```diff
@@ -106,52 +104,54 @@ class RemoveUndef : public IRMutator {
...
         }
     }
 
-    void visit(const Broadcast *op) {
+    Expr visit(const Broadcast *op) override {
         Expr value = mutate(op->value);
-        if (!value.defined()) return;
-        if (value.same_as(op->value)) expr = op;
-        else expr = Broadcast::make(value, op->lanes);
+        if (!value.defined()) return op;
```
Comment (abadams) [inline src/RemoveUndef.cpp]: return Expr()

File: src/RemoveUndef.cpp [hunk 7]
```diff
@@ -183,19 +183,20 @@ class RemoveUndef : public IRMutator {
         if (!value.defined()) {
             dead_vars.pop(op->name);
         }
-        if (!body.defined()) return;
+        if (!body.defined()) return op;
```
Comment (abadams) [inline src/RemoveUndef.cpp]: return Expr()

File: src/RemoveUndef.cpp [hunk 8]
```diff
@@ -204,99 +205,94 @@ class RemoveUndef : public IRMutator {
...
```
Comment (abadams) [inline src/RemoveUndef.cpp]: return Expr()

File: src/RemoveUndef.cpp [hunk 9]
```diff
@@ -204,99 +205,94 @@ class RemoveUndef : public IRMutator {
...
-            stmt = AssertStmt::make(condition, message);
+            return AssertStmt::make(condition, message);
         }
     }
 
-    void visit(const ProducerConsumer *op) {
+    Stmt visit(const ProducerConsumer *op) override {
         Stmt body = mutate(op->body);
-        if (!body.defined()) return;
+        if (!body.defined()) return op;
```
Comment (abadams) [inline src/RemoveUndef.cpp]: return Expr()

File: src/RemoveUndef.cpp [hunk 10]
```diff
@@ -204,99 +205,94 @@ class RemoveUndef : public IRMutator {
...
         }
         Expr extent = mutate(op->extent);
         if (!extent.defined()) {
-            stmt = Stmt();
-            return;
+            return Stmt();
         }
         Stmt body = mutate(op->body);
-        if (!body.defined()) return;
+        if (!body.defined()) return op;
```
Comment (abadams) [inline src/RemoveUndef.cpp]: return Expr()

File: src/RemoveUndef.cpp [hunk 11]
```diff
@@ -354,31 +348,35 @@ class RemoveUndef : public IRMutator {
...
-                stmt = Stmt();
-                return;
+                return Stmt();
             }
             all_extents_unmodified &= new_extents[i].same_as(op->extents[i]);
         }
         Stmt body = mutate(op->body);
-        if (!body.defined()) return;
+        if (!body.defined()) {
+            return op;
```
Comment (abadams) [inline src/RemoveUndef.cpp]: return Expr()

File: src/RemoveUndef.cpp [hunk 12]
```diff
@@ -354,31 +348,35 @@ class RemoveUndef : public IRMutator {
...
+        }
 
         Expr condition = mutate(op->condition);
-        if (!condition.defined()) return;
+        if (!condition.defined()) {
+            return op;
```
Comment (abadams) [inline src/RemoveUndef.cpp]: return Expr()

File: src/RemoveUndef.cpp [hunk 13]
```diff
@@ -409,61 +407,61 @@ class RemoveUndef : public IRMutator {
...
             }
             if (!new_min.same_as(old_min))       bounds_changed = true;
             if (!new_extent.same_as(old_extent)) bounds_changed = true;
             new_bounds[i] = Range(new_min, new_extent);
         }
 
         Stmt body = mutate(op->body);
-        if (!body.defined()) return;
+        if (!body.defined()) {
+            return op;
```
Comment (abadams) [inline src/RemoveUndef.cpp]: return Expr()

File: src/RemoveUndef.cpp (hunk 1 above)
Comment (steven-johnson) [reply src/RemoveUndef.cpp]: To mimic existing behavior, it should return whatever the default value of this->expr is... which (looks at code) is undefined. So, yeah.

File: src/RemoveUndef.cpp [hunk 14]
```diff
@@ -409,61 +407,61 @@ class RemoveUndef : public IRMutator {
...
+        }
 
         Expr condition = mutate(op->condition);
-        if (!condition.defined()) return;
+        if (!condition.defined()) {
+            return op;
```
Comment (abadams) [inline src/RemoveUndef.cpp]: return Expr()

**Comment by steven-johnson** [comment]
PTAL

**Comment by abadams** [comment]
I think you missed a few - most of my comments about went away due to changes in the branch, but three remain.

File: src/RemoveUndef.cpp (hunk 8 above)
Comment (steven-johnson) [reply src/RemoveUndef.cpp]: actually, return Stmt()

**Comment by steven-johnson** [comment]
I only saw two remaining.

**Comment by abadams** [comment]
Did another scan - I'm pretty sure you got them all. LGTM.

**Pull request merged**